File: PruebaMitM.py
#!/usr/bin/env python3
"""
PruebaMitM.py

Prueba de concepto (PoC) de ataque Man-in-the-Middle (MitM) contra la VPN SSL BYOD.

Este script:
  1. Genera un certificado SSL autofirmado FALSO (diferente al del servidor real).
  2. Levanta un proxy TLS en un puerto intermedio (4443).
  3. Cuando el cliente Java intenta conectar al proxy, el handshake TLS FALLA
     porque el truststore del cliente solo confía en el certificado del servidor real.
  4. Demuestra que la configuración estricta del TrustStore mitiga ataques MitM.

Uso:
  1. Asegurar que el servidor real está corriendo en puerto 3443.
  2. Ejecutar este script:
       python PruebaMitM.py
  3. En otra terminal, intentar conectar el cliente al proxy (puerto 4443):
       java -Djavax.net.ssl.trustStore=cliente_truststore.jks 
            -Djavax.net.ssl.trustStorePassword=cambiame 
            -Djavax.net.ssl.host=localhost 
            ClienteSSL
     (Modificar temporalmente ClienteSSL.java para conectar al puerto 4443,
      o usar la versión de test que acepta el puerto por parámetro.)

Resultado esperado:
  - El cliente Java lanzará javax.net.ssl.SSLHandshakeException
  - El proxy MitM NO podrá interceptar la comunicación
  - Esto demuestra que TLS 1.3 + TrustStore previene MitM

Requisitos:
  - Python 3.6+
  - No se necesitan dependencias externas

@author Manuel
"""

import ssl
import socket
import threading
import subprocess
import os
import sys
import time
import tempfile

# ======================== CONFIGURACIÓN ========================

PUERTO_PROXY = 4443            # Puerto donde escucha el proxy MitM
SERVIDOR_REAL_HOST = "localhost"
SERVIDOR_REAL_PUERTO = 3443    # Puerto del servidor SSL real
CERT_FALSO = "mitm_cert.pem"   # Certificado falso generado
KEY_FALSA = "mitm_key.pem"     # Clave privada falsa generada


def generar_certificado_falso():
    """
    Genera un certificado SSL autofirmado FALSO para el proxy MitM.
    Usa el módulo ssl de Python para generar un par clave/certificado.
    El CN será 'localhost' para intentar engañar al cliente,
    pero el certificado NO estará en el truststore del cliente Java.
    """
    print("[MitM] Generando certificado falso autofirmado...")
    
    # Usar openssl para generar el certificado (disponible en la mayoría de sistemas)
    try:
        # Generar clave privada RSA 2048
        subprocess.run([
            "openssl", "req", "-x509", "-newkey", "rsa:2048",
            "-keyout", KEY_FALSA,
            "-out", CERT_FALSO,
            "-days", "1",
            "-nodes",  # Sin contraseña en la clave
            "-subj", "/CN=localhost/O=MitM_Atacante/C=XX"
        ], check=True, capture_output=True, text=True)
        
        print(f"[MitM] Certificado falso generado: {CERT_FALSO}")
        print(f"[MitM] Clave falsa generada: {KEY_FALSA}")
        print(f"[MitM] CN=localhost, O=MitM_Atacante (NO coincide con el truststore del cliente)")
        return True
        
    except FileNotFoundError:
        print("[ERROR] OpenSSL no encontrado. Intentando método alternativo con keytool...")
        return generar_certificado_con_keytool()
    except subprocess.CalledProcessError as e:
        print(f"[ERROR] Error generando certificado: {e.stderr}")
        return False


def generar_certificado_con_keytool():
    """
    Método alternativo usando keytool de Java para generar certificado falso.
    """
    try:
        # Generar keystore falso
        subprocess.run([
            "keytool", "-genkeypair",
            "-alias", "mitm",
            "-keyalg", "RSA",
            "-keysize", "2048",
            "-validity", "1",
            "-keystore", "mitm_keystore.jks",
            "-storepass", "atacante",
            "-dname", "CN=localhost, OU=MitM, O=Atacante, C=XX"
        ], check=True, capture_output=True, text=True)
        
        # Exportar a PEM (necesita openssl, así que simplemente usamos el JKS directamente)
        print("[MitM] Keystore MitM generado: mitm_keystore.jks")
        return True
        
    except Exception as e:
        print(f"[ERROR] No se pudo generar certificado: {e}")
        return False


def proxy_mitm():
    """
    Levanta un servidor TLS proxy que intenta interceptar la comunicación.
    
    Funcionamiento:
    1. El proxy acepta conexiones TLS en PUERTO_PROXY con su certificado FALSO.
    2. Si un cliente acepta el certificado falso (MAL CONFIGURADO), el proxy
       reenvía el tráfico al servidor real en SERVIDOR_REAL_PUERTO.
    3. Si el cliente RECHAZA el certificado (bien configurado con TrustStore),
       el handshake falla y la conexión se cierra. Esto es lo ESPERADO.
    """
    print(f"\n[MitM] ═══════════════════════════════════════════════════")
    print(f"[MitM]   PROXY MAN-IN-THE-MIDDLE INICIADO")
    print(f"[MitM]   Escuchando en puerto {PUERTO_PROXY}")
    print(f"[MitM]   Servidor real en {SERVIDOR_REAL_HOST}:{SERVIDOR_REAL_PUERTO}")
    print(f"[MitM] ═══════════════════════════════════════════════════\n")
    
    # Crear contexto SSL con certificado falso
    contexto = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
    
    try:
        contexto.load_cert_chain(certfile=CERT_FALSO, keyfile=KEY_FALSA)
    except Exception as e:
        print(f"[ERROR] No se pudo cargar el certificado falso: {e}")
        return
    
    # Permitir TLS 1.3
    contexto.minimum_version = ssl.TLSVersion.TLSv1_3
    
    # Crear socket servidor
    servidor = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    servidor.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    servidor.bind(("0.0.0.0", PUERTO_PROXY))
    servidor.listen(5)
    
    print(f"[MitM] Esperando conexiones de víctimas en puerto {PUERTO_PROXY}...")
    print(f"[MitM] Para probar, ejecute el cliente Java apuntando al puerto {PUERTO_PROXY}:\n")
    print(f"  Modifique HOST/PUERTO en ClienteSSL.java o ejecute:")
    print(f"    java -Djavax.net.ssl.trustStore=cliente_truststore.jks \\")
    print(f"         -Djavax.net.ssl.trustStorePassword=cambiame \\")
    print(f"         ClienteSSLMitM\n")
    
    intentos = 0
    while True:
        try:
            conn_raw, addr = servidor.accept()
            intentos += 1
            print(f"\n[MitM] ¡Conexión entrante #{intentos} desde {addr[0]}:{addr[1]}!")
            print(f"[MitM] Intentando handshake TLS con certificado FALSO...")
            
            try:
                conn_ssl = contexto.wrap_socket(conn_raw, server_side=True)
                
                # Si llegamos aquí, el cliente ACEPTÓ el certificado falso (MAL)
                print(f"[MitM] ⚠️  ¡ALERTA! El cliente ACEPTÓ el certificado falso.")
                print(f"[MitM] Esto significa que el TrustStore NO está configurado correctamente.")
                print(f"[MitM] Interceptando tráfico...")
                
                # Leer datos del cliente
                datos = conn_ssl.recv(4096)
                print(f"[MitM] Datos interceptados: {datos.decode('utf-8', errors='replace')}")
                
                conn_ssl.close()
                
            except ssl.SSLError as e:
                print(f"[MitM] ✅ Handshake FALLIDO (esperado): {e}")
                print(f"[MitM] ✅ El cliente Java RECHAZÓ el certificado falso.")
                print(f"[MitM] ✅ ¡El TrustStore protege contra este ataque MitM!")
                
            except ConnectionResetError:
                print(f"[MitM] ✅ Conexión reseteada por el cliente (rechazó el cert).")
                print(f"[MitM] ✅ ¡Protección MitM funcionando correctamente!")
                
            except Exception as e:
                print(f"[MitM] Error en la conexión: {type(e).__name__}: {e}")
                
        except KeyboardInterrupt:
            print("\n[MitM] Proxy detenido por el usuario.")
            break
    
    servidor.close()
    limpiar_archivos()


def limpiar_archivos():
    """Elimina los archivos temporales del certificado falso."""
    for archivo in [CERT_FALSO, KEY_FALSA]:
        if os.path.exists(archivo):
            os.remove(archivo)
            print(f"[MitM] Eliminado: {archivo}")


def crear_cliente_test_mitm():
    """
    Genera un fichero Java auxiliar (ClienteSSLMitM.java) que conecta
    al puerto del proxy MitM (4443) en lugar del servidor real.
    Esto permite probar el ataque sin modificar ClienteSSL.java.
    """
    codigo = '''import java.io.*;
import javax.net.ssl.*;

/**
 * ClienteSSLMitM.java
 * 
 * Cliente auxiliar que intenta conectar al proxy MitM (puerto 4443).
 * Se espera que el handshake TLS FALLE porque el certificado del proxy
 * no está en el truststore del cliente.
 * 
 * Ejecución:
 *   java -Djavax.net.ssl.trustStore=cliente_truststore.jks 
 *        -Djavax.net.ssl.trustStorePassword=cambiame 
 *        ClienteSSLMitM
 */
public class ClienteSSLMitM {
    public static void main(String[] args) {
        System.out.println("=== Prueba MitM: Conectando al proxy atacante (puerto 4443) ===\\n");
        
        try {
            SSLSocketFactory factory = (SSLSocketFactory) SSLSocketFactory.getDefault();
            SSLSocket socket = (SSLSocket) factory.createSocket("localhost", 4443);
            socket.setEnabledProtocols(new String[]{"TLSv1.3"});
            
            System.out.println("Intentando handshake TLS con el proxy MitM...");
            socket.startHandshake();
            
            // Si llegamos aquí, algo va MAL
            System.out.println("⚠️  ALERTA: El handshake fue exitoso con el proxy atacante.");
            System.out.println("El TrustStore NO está protegiendo correctamente.");
            socket.close();
            
        } catch (javax.net.ssl.SSLHandshakeException e) {
            System.out.println("✅ SSLHandshakeException capturada correctamente.");
            System.out.println("✅ El cliente RECHAZÓ el certificado falso del atacante.");
            System.out.println("✅ El TrustStore protege contra el ataque MitM.");
            System.out.println("\\nDetalle de la excepción:");
            System.out.println("  " + e.getMessage());
            
        } catch (Exception e) {
            System.out.println("Error: " + e.getClass().getSimpleName() + ": " + e.getMessage());
        }
    }
}
'''
    
    with open("ClienteSSLMitM.java", "w", encoding="utf-8") as f:
        f.write(codigo)
    
    print(f"[MitM] Archivo auxiliar generado: ClienteSSLMitM.java")
    print(f"[MitM] Compílelo con: javac ClienteSSLMitM.java")


# ======================== MAIN ========================

if __name__ == "__main__":
    print("╔══════════════════════════════════════════════════════╗")
    print("║   VPN SSL BYOD - Prueba de Concepto MitM            ║")
    print("║   Man-in-the-Middle Attack Simulation               ║")
    print("╚══════════════════════════════════════════════════════╝\n")
    
    print("OBJETIVO: Demostrar que TLS 1.3 + TrustStore previene")
    print("la interceptación de tráfico por un atacante.\n")
    
    # Paso 1: Generar certificado falso
    if not generar_certificado_falso():
        print("[ERROR] No se pudo generar el certificado MitM. Abortando.")
        sys.exit(1)
    
    # Paso 2: Generar cliente auxiliar de test
    crear_cliente_test_mitm()
    
    # Paso 3: Iniciar proxy MitM
    print("\n" + "="*60)
    print("INSTRUCCIONES:")
    print("  1. Deje este script corriendo (es el proxy MitM).")
    print("  2. En OTRA terminal, compile y ejecute el cliente MitM:")
    print(f"     javac ClienteSSLMitM.java")
    print(f"     java -Djavax.net.ssl.trustStore=cliente_truststore.jks \\")
    print(f"          -Djavax.net.ssl.trustStorePassword=cambiame ClienteSSLMitM")
    print("  3. Observe cómo el cliente RECHAZA la conexión.")
    print("  4. Presione Ctrl+C para detener el proxy.")
    print("="*60 + "\n")
    
    proxy_mitm()

File: test/PruebaMitM.java
import java.io.*;
import java.net.*;
import java.security.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import javax.net.ssl.*;

/**
 * PruebaMitM.java
 * 
 * Prueba de concepto (PoC) de ataque Man-in-the-Middle contra la VPN SSL BYOD.
 * 
 * Este programa:
 * 1. Genera un keystore con un certificado autofirmado FALSO en memoria.
 * 2. Levanta un proxy TLS en el puerto 4443.
 * 3. Cuando el cliente Java intenta conectar, el handshake TLS FALLA
 * porque el truststore del cliente solo confía en el certificado real del
 * servidor.
 * 4. Demuestra que la configuración estricta del TrustStore mitiga ataques
 * MitM.
 * 
 * Ejecución:
 * 1. Arrancar el servidor real:
 * java -cp ".;sqlite-jdbc-3.47.2.0.jar"
 * -Djavax.net.ssl.keyStore=servidor_keystore.jks
 * -Djavax.net.ssl.keyStorePassword=cambiame ServidorSSL
 * 
 * 2. Arrancar este proxy MitM:
 * java PruebaMitM
 * 
 * 3. Ejecutar el cliente apuntando al proxy (puerto 4443):
 * java -Djavax.net.ssl.trustStore=cliente_truststore.jks
 * -Djavax.net.ssl.trustStorePassword=cambiame ClienteSSLMitM
 * 
 * Resultado esperado:
 * - El cliente lanza SSLHandshakeException → conexión RECHAZADA ✅
 * - El proxy NO puede interceptar la comunicación
 * 
 * @author Manuel
 */
public class PruebaMitM {

    /** Puerto donde escucha el proxy MitM */
    private static final int PUERTO_PROXY = 4443;

    /** Puerto del servidor real */
    private static final int PUERTO_SERVIDOR_REAL = 3443;

    /** Alias y contraseña del keystore MitM generado en memoria */
    private static final String ALIAS_MITM = "mitm_atacante";
    private static final String PASSWORD_MITM = "atacante";

    // ======================== MAIN ========================

    public static void main(String[] args) {
        System.out.println("╔══════════════════════════════════════════════════════╗");
        System.out.println("║   VPN SSL BYOD - Prueba de Concepto MitM (Java)     ║");
        System.out.println("║   Man-in-the-Middle Attack Simulation               ║");
        System.out.println("╚══════════════════════════════════════════════════════╝\n");

        System.out.println("OBJETIVO: Demostrar que TLS 1.3 + TrustStore previene");
        System.out.println("la interceptación de tráfico por un atacante.\n");

        try {
            // Paso 1: Generar keystore MitM con certificado falso
            System.out.println("[MitM] Paso 1: Generando certificado falso...");
            KeyStore keystoreMitM = generarKeystoreFalso();
            System.out.println("[MitM] Certificado falso generado en memoria.");
            System.out.println("[MitM]   CN=localhost, O=MitM_Atacante, C=XX");
            System.out.println("[MitM]   (NO está en el truststore del cliente)\n");

            // Paso 2: Configurar SSLContext con el keystore falso
            SSLContext contextoMitM = SSLContext.getInstance("TLSv1.3");
            KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
            kmf.init(keystoreMitM, PASSWORD_MITM.toCharArray());
            contextoMitM.init(kmf.getKeyManagers(), null, new SecureRandom());

            // Paso 3: Crear SSLServerSocket del proxy
            SSLServerSocketFactory ssf = contextoMitM.getServerSocketFactory();
            SSLServerSocket proxySocket = (SSLServerSocket) ssf.createServerSocket(PUERTO_PROXY);
            proxySocket.setEnabledProtocols(new String[] { "TLSv1.3" });

            System.out.println("[MitM] ═══════════════════════════════════════════════");
            System.out.println("[MitM]   PROXY MAN-IN-THE-MIDDLE INICIADO");
            System.out.println("[MitM]   Escuchando en puerto " + PUERTO_PROXY);
            System.out.println("[MitM]   Servidor real en localhost:" + PUERTO_SERVIDOR_REAL);
            System.out.println("[MitM] ═══════════════════════════════════════════════\n");

            System.out.println("[MitM] Esperando conexiones de víctimas...");
            System.out.println("[MitM] Execute en otra terminal:");
            System.out.println("         java -Djavax.net.ssl.trustStore=cliente_truststore.jks \\");
            System.out.println("              -Djavax.net.ssl.trustStorePassword=cambiame ClienteSSLMitM\n");

            // Crear fichero de log
            new File("logs").mkdirs();
            String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd_HH-mm-ss"));
            String nombreLog = "logs/PruebaMitM_" + timestamp + ".log";
            PrintWriter log = new PrintWriter(new FileWriter(nombreLog), true);
            log.println("========================================");
            log.println("  TEST: PruebaMitM");
            log.println("  FECHA: " + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")));
            log.println("========================================");
            log.println();
            System.out.println("[LOG] Registrando resultados en: " + nombreLog + "\n");

            int intentos = 0;
            while (true) {
                try {
                    Socket clienteRaw = proxySocket.accept();
                    intentos++;
                    String clienteIP = clienteRaw.getInetAddress().getHostAddress();

                    System.out.println("\n[MitM] ¡Conexión entrante #" + intentos + " desde " + clienteIP + "!");
                    System.out.println("[MitM] Intentando handshake TLS con certificado FALSO...");

                    // Intentar leer datos (el handshake ya ocurrió en accept para SSLSocket)
                    BufferedReader entrada = new BufferedReader(
                            new InputStreamReader(clienteRaw.getInputStream(), "UTF-8"));

                    // Si llegamos aquí, el cliente ACEPTÓ el certificado falso (MAL)
                    System.out.println("[MitM]  ¡ALERTA! El cliente ACEPTÓ el certificado falso.");
                    System.out.println("[MitM] El TrustStore NO está configurado correctamente.");

                    String datos = entrada.readLine();
                    if (datos != null) {
                        System.out.println("[MitM] Datos interceptados: " + datos);
                    }

                    log.println("[" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")) + "] Intento #" + intentos + " desde " + clienteIP + ": ALERTA - Cliente ACEPTÓ certificado falso");
                    clienteRaw.close();

                } catch (SSLHandshakeException e) {
                    System.out.println("[MitM] Handshake TLS FALLIDO (resultado esperado).");
                    System.out.println("[MitM] El cliente Java RECHAZÓ el certificado falso.");
                    System.out.println("[MitM] ¡El TrustStore protege contra este ataque MitM!");
                    System.out.println("[MitM]    Detalle: " + e.getMessage());
                    log.println("[" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")) + "] Intento #" + (intentos + 1) + ": Handshake FALLIDO (protección MitM OK) - " + e.getMessage());

                } catch (IOException e) {
                    System.out.println("[MitM] Conexión rechazada/reseteada por el cliente.");
                    System.out.println("[MitM] ¡Protección MitM funcionando correctamente!");
                    System.out.println("[MitM]    Detalle: " + e.getClass().getSimpleName() + ": " + e.getMessage());
                    log.println("[" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")) + "] Intento #" + (intentos + 1) + ": Conexión rechazada (protección MitM OK) - " + e.getClass().getSimpleName() + ": " + e.getMessage());
                }
            }

        } catch (BindException e) {
            System.err.println("[ERROR] El puerto " + PUERTO_PROXY + " ya está en uso.");
        } catch (Exception e) {
            System.err.println("[ERROR] Error al iniciar el proxy MitM: " + e.getMessage());
            e.printStackTrace();
        }
    }

    // ======================== GENERACIÓN DE KEYSTORE FALSO
    // ========================

    /**
     * Genera un KeyStore en memoria con un certificado autofirmado FALSO.
     * Usa keytool internamente a través de la API de Java.
     * 
     * El certificado tiene:
     * - CN=localhost (para intentar engañar al cliente)
     * - O=MitM_Atacante (organización falsa)
     * - Validez de 1 día
     * - RSA 2048 bits
     * 
     * Este certificado NO está en el truststore del cliente, por lo que
     * el handshake TLS será rechazado.
     */
    private static KeyStore generarKeystoreFalso() throws Exception {
        String keystorePath = "mitm_keystore_temp.jks";

        // Eliminar keystore anterior si existe
        new File(keystorePath).delete();

        // Usar keytool para generar el keystore falso
        ProcessBuilder pb = new ProcessBuilder(
                "keytool", "-genkeypair",
                "-alias", ALIAS_MITM,
                "-keyalg", "RSA",
                "-keysize", "2048",
                "-validity", "1",
                "-keystore", keystorePath,
                "-storepass", PASSWORD_MITM,
                "-keypass", PASSWORD_MITM,
                "-dname", "CN=localhost, OU=MitM, O=MitM_Atacante, L=Desconocido, ST=Desconocido, C=XX");
        pb.redirectErrorStream(true);
        Process proceso = pb.start();

        // Consumir salida del proceso
        BufferedReader reader = new BufferedReader(new InputStreamReader(proceso.getInputStream()));
        while (reader.readLine() != null) {
            // Ignorar salida de keytool
        }
        int exitCode = proceso.waitFor();

        if (exitCode != 0) {
            throw new RuntimeException("Error al generar keystore MitM con keytool (exit code: " + exitCode + ")");
        }

        // Cargar el keystore generado
        KeyStore ks = KeyStore.getInstance("JKS");
        try (FileInputStream fis = new FileInputStream(keystorePath)) {
            ks.load(fis, PASSWORD_MITM.toCharArray());
        }

        System.out.println("[MitM] Keystore MitM generado: " + keystorePath);
        return ks;
    }
}
